client/network/src/gossip.rs

```diff
@@ -0,0 +1,318 @@
+//! the notifications will be interleaved in an unpredictable way.
+//!
+
+use crate::{ExHashT, NetworkService, service::{NotificationSender, NotificationSenderError}};
+
+use async_std::sync::{Condvar, Mutex, MutexGuard};
+use futures::prelude::*;
+use libp2p::PeerId;
+use sp_runtime::{traits::Block as BlockT, ConsensusEngineId};
+use std::{
+	collections::VecDeque,
+	fmt,
+	sync::{atomic, Arc},
+	time::Duration,
+};
+
+/// Notifications sender for a specific combination of network service, peer, and protocol.
+pub struct DirectedGossip<M> {
+	/// Shared between the front and the back task.
+	shared: Arc<Shared<M>>,
+}
+
+impl<M> DirectedGossip<M> {
+	/// Returns a new [`DirectedGossip`] containing a queue of message for this specific
+	/// combination of peer and protocol.
+	///
+	/// In addition to the [`DirectedGossip`], also returns a `Future` whose role is to drive
+	/// the messages sending forward.
+	pub fn new<B, H, F>(
+		service: Arc<NetworkService<B, H>>,
+		peer_id: PeerId,
+		protocol: ConsensusEngineId,
+		queue_size_limit: usize,
+		messages_encode: F
+	) -> (Self, impl Future<Output = ()> + Send + 'static)
+	where
+		M: Send + 'static,
+		B: BlockT + 'static,
+		H: ExHashT,
+		F: Fn(M) -> Vec<u8> + Send + 'static,
+	{
+		DirectedGossipPrototype::new(service, peer_id)
+			.build(protocol, queue_size_limit, messages_encode)
+	}
+
+	/// Locks the queue of messages towards this peer.
+	///
+	/// The returned `Future` is expected to be ready quite quickly.
+	pub async fn lock_queue<'a>(&'a self) -> QueueLock<'a, M> {
+		QueueLock {
+			lock: self.shared.locked.lock().await,
+			condvar: &self.shared.condvar,
+			queue_size_limit: self.shared.queue_size_limit,
+		}
+	}
+
+	/// Pushes a message to the queue, or discards it if the queue is full.
+	///
+	/// The returned `Future` is expected to be ready quite quickly.
+	pub async fn queue_or_discard(&self, message: M)
+	where
+		M: Send + 'static
+	{
+		self.lock_queue().await.push_or_discard(message);
+	}
+}
+
+impl<M> fmt::Debug for DirectedGossip<M> {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_struct("DirectedGossip").finish()
+	}
+}
+
+impl<M> Drop for DirectedGossip<M> {
+	fn drop(&mut self) {
+		// The "clean" way to notify the `Condvar` here is normally to first lock the `Mutex`,
+		// then notify the `Condvar` while the `Mutex` is locked. Unfortunately, the `Mutex`
+		// being asynchronous, it can't reasonably be locked from within a destructor.
+		// For this reason, this destructor is a "best effort" destructor.
+		// See also the corresponding code in the background task.
+		self.shared.stop_task.store(true, atomic::Ordering::Acquire);
+		self.shared.condvar.notify_all();
+	}
+}
+
+/// Utility. Generic over the type of the messages. Holds a [`NetworkService`] and a [`PeerId`].
+/// Provides a [`DirectedGossipPrototype::build`] function that builds a [`DirectedGossip`].
+#[derive(Clone)]
+pub struct DirectedGossipPrototype {
```

Comment:
This is `pub` so I assume this will be used from somewhere external? Could you explain this a little more please? Currently it is not in use and the `DirectedGossip` and task construction could happen directly in `DirectedGossip::new`.

Reply:
I admit that this is type is specifically targeted for the Polkadot use case.

In Polkadot's code, the so-called network bridge communicates [via messages](https://github.com/paritytech/polkadot/blob/cc2d7afd128b259791139359067aa519eadbf205/node/subsystem/src/messages.rs#L138-L153) with other subsystems. For example, when we connect, a `PeerConnected` message is sent.

The network bridge doesn't know what is the type of the messages that the various subsystems would manipulate (the `M` generics), and the various subsystems don't know what the `B` and `H` generics would be.

This `DirectedGossipPrototype` would easily plug into that scheme.

Reply:
Should Polkadot-specific code not be put into Polkadot and not in here?

Reply:
> The network bridge doesn't know what is the type of the networking messages that the various subsystems would manipulate (the M generic in this PR), and the various subsystems don't know what the B and H generics would be.

I am actually changing this right now, so we may not need this type. Will push a PR shortly

Reply:
https://github.com/paritytech/polkadot/pull/1535 - now the network bridge is aware of the specific information flowing over the network and we may be able to avoid these prototypes. However, different subsystems will still want `Sender`s for different types. However, the sender for a specific variant should also be encoding as `|x| ValidationProtocolV1::SomeConcreteVariant(x).encode()` which I think is covered by the encode-fn part of the API here

Reply:
> Should Polkadot-specific code not be put in Polkadot?

To answer specifically this: the way I see it, this code isn't really Polkadot-specific.
This code provides a convenient-to-use high-level API on top of a low-level primitive, and people are free to use either the convenient API if it suits them, or the low-level primitive if it doesn't suit them.

I think that, in general, we want to be as "universal" as possible when designing low-level code, but as long as the lower-level primitives are exposed, higher-level code can be a bit more targeted towards certain use-cases.